<?xml version="1.0"?>
<launch>
  
  <arg name="serial" default="150534" />
  <arg name="rqt" default="True" />
  <arg name="rviz" default="False" />
  
  <!-- Logging -->
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find ensenso)/config/rosconsole.conf"/>
  
  <!-- Ensenso driver -->
  <node name="ensenso_driver" pkg="ensenso" type="ensenso_driver" ns="camera" output="screen"> 
    <param name="serial"          type="string" value="$(arg serial)" />
    <param name="camera_frame_id" type="string" value="ensenso_optical_frame" />
  </node>
  
  <!-- Driver configuration -->
  <node name="driver_dynconfig" pkg="dynamic_reconfigure" type="dynparam" args="set_from_parameters ensenso_driver" ns="camera">
    <!-- Capture parameters -->
    <param name="AutoBlackLevel"    type="bool"   value="true" />
    <param name="AutoExposure"      type="bool"   value="true" />
    <param name="AutoGain"          type="bool"   value="true" />
    <param name="Binning"           type="int"    value="1" />
    <param name="BlackLevelOffset"  type="double" value="1.0" />
    <param name="DisparityMapAOI"   type="bool"   value="false" />
    <param name="Exposure"          type="double" value="1.5" />
    <param name="FlexView"          type="bool"   value="false" />
    <param name="FlexViewImages"    type="int"    value="2" />
    <param name="FrontLight"        type="bool"   value="false" />
    <param name="Gain"              type="int"    value="1" />
    <param name="GainBoost"         type="bool"   value="true" />
    <param name="HardwareGamma"     type="bool"   value="true" />
    <param name="Hdr"               type="bool"   value="true" />
    <param name="PixelClock"        type="int"    value="24" />
    <param name="Projector"         type="bool"   value="true" />
    <param name="TargetBrightness"  type="int"    value="80" />
    <param name="TriggerMode"       type="int"    value="0" />      <!-- Software: 0, FallingEdge: 1, RisingEdge: 2 -->
    <!-- Stereo parameters -->
    <param name="MinimumDisparity"      type="int"    value="-117" />
    <param name="NumberOfDisparities"   type="int"    value="64" />
    <param name="OptimizationProfile"   type="int"    value="2" />  <!-- Aligned: 0, Diagonal: 1, AlignedAndDiagonal: 2 -->
    <param name="Scaling"               type="double" value="1.0" />
    <!-- Streaming parameters -->    
    <param name="Cloud"   type="bool"   value="true" />
    <param name="Images"  type="bool"   value="true" />
  </node>
  
  <group if="$(arg rqt)">
    <node name="rqt_gui_node" pkg="rqt_gui" type="rqt_gui" args="--perspective-file $(find ensenso)/config/viewer.perspective" />
  </group>
  
  <group if="$(arg rviz)">
    <node name="rviz_node" pkg="rviz" type="rviz" args="-d $(find ensenso)/config/viewer.rviz" />
  </group>

</launch>
